[preview] Openpulse assembler

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Fake PR to clarify the diff


### Details and comments

itoko committed:
make commands callable

itoko committed:
make channels hashable

itoko committed:
add working example code of Schedule

itoko committed:
reorder class defs and add prints

itoko committed:
simplify Schedule

itoko committed:
rename PulseChannel->Channel

itoko committed:
lint

itoko committed:
add getters for qobj creation

itoko committed:
simplify snapshot command

itoko committed:
rename Pulse -> Instruction

itoko committed:
simplify Instruction

itoko committed:
refactor schedule-related interface

itoko committed:
unify exceptions into PulseError

itoko committed:
wip: make schedule immutable

itoko committed:
make schedule immutable

itoko committed:
fix a bug in copying of a Schedule

itoko committed:
add check length of mem_slots

itoko committed:
make sample pulses hashable

itoko committed:
fix a bug in Acquire constructor with default values

itoko committed:
refactor meas_opts

Commit:
remove all changes except for commands-rework

Commit:
[wip] apply updates for assembler

Commit:
[wip] apply updates for assembler

1ucian0 committed:
fixed_point in the same style as dag_fixed_point (#2019)

nkanazawa1989 commented:
Do not merge this

diego-plan9 committed:
Remove underscores from qiskit.qasm (#2080)

* Remove underscore from files at qiskit.qasm

* Update module docstrings style

* Replace super calls to node subclasses

mtreinish committed:
Remove underscores from interactive visualization modules (#2081)

* Remove underscores from interactive visualization modules

To be consistent with the rest of the visualization modules this commit
removes the leading underscore from the interactive visualization
modules.

* Fix missed imports

mtreinish committed:
Make test running concurrency dynamic based on nprocs (#1973)

* Make test running concurrency dynamic based on nprocs

In #1228 we decreased the concurrency used for the test runner in CI to
avoid memory pressure and failing jobs. In that commit we set it to 2 as
a quick fix so we could unblock CI and continue landing patches. A
followup was to do the analysis and find the sweet spot for concurrency
and stability. At the same time we opened mtreinish/stestr#202 to add
support to the test runner for running fractional nprocs. That is still
under discussion, but in the meantime this commit adds support to the
makefile for using fractional nprocs concurrency. It adjusts are default
concurrency from 2 to be nprocs / 2 if nprocs > 3. This should enable us
to increase the number of test workers we run on travis linux jobs but
still maintain reliability.

* Hard code osx to 2 workers

* Add debug statement about core count and worker count

* Use printf to round floats to int

chriseclectic committed:
Remove `inplace` from QuantumChannel and Unitary (#2077)

* remove inplace kwarg from QuantumChannel

* remove inplace channel tests

* remove inplace from Unitary

* reduce power

prakashmurali committed:
Noise-adaptive initial qubit layout pass (#2089)

* Added a noise-adaptive initial layout pass. Maps virtual qubits using backend calibration data to maximize program success rate

* CHANGELOG update (2089)

* Fixed __init__.py in transpiler passes

* Update qiskit/transpiler/passes/mapping/noise_adaptive_layout.py

Co-Authored-By: prakashmurali <prakashmurali@gmail.com>

* Update qiskit/transpiler/passes/mapping/noise_adaptive_layout.py

Co-Authored-By: prakashmurali <prakashmurali@gmail.com>

* Updated qreg to qarg

taalexander committed:
Add meas_level to result schema (#2085)

* Add meas_level to result schema.

* Update changelog.

* Added meas_level to marshmallow schema.

* Remove required from result model.

1ucian0 committed:
Extend and enlarge the the dag/layout with ancillas using the full coupling map (#2023)

* Extend and enlarge the the dag/layout with ancillas using the full coupling map

* unused import

* test_final_measurement_barrier_for_devices

* test_move_measurements

* klint

* unused import

* add tests

* changelog

1ucian0 committed:
Optimize gates up to fixed point in the default pipeline (#2086)

* Optimze gates up to fixed point in the default pipeline

* import

* allow transpile_dag to be called with initial_layout=None

itoko committed:
lint

itoko committed:
make mem_slots independent of qubits

eggerdj commented:
Hi Itoko, I was reading through the code and have a question: what is the rational to have the LO in the schedules?  I am referring in particular to lines 163 to 175 of assembler.py. For instance assembling schedules with different LOs could lead to unexpected behavior or pulses that the hardware cannot realize?

ajavadia committed:
import unitary in quantum_info package (#2093)

itoko commented:
@eggerdj Thank you for your comment. Could you look at `execute.py`? . You can specify 1 `Schedule` with n LO settings or n `Schedule`s with 1 LO setting in `execute_schedule()`. (Excuse me for naming, `ConfiguredSchedule` is not a `Schedule`). Am I understanding your concern correctly?
```
[Proposed (1 schedule : n lo_configs)]
# compose schedule
schedule = pulse.Schedule()
#... compose schedule ...

# set LO frequency
lo_configs=[]
for drive_lo_freq in drive_lo_freqs:
    freq_d0 = freq_d0_default + drive_lo_freq
    lo_configs.append(ScheduleConfig(user_lo_dic={device.q[0].drive: freq_d0}))

execute_schedules(schedule, backend, lo_configs, **opts)
```
```
[Proposed (n schedules : 1 lo_config)]
# compose to schedules
schedules = []
for ii in pulses:
    #... compose ii-th schedule ...
    schedules.append(schedule)

# set LO frequency
lo_config = ScheduleConfig(user_lo_dic={device.q[0].drive: freq_d0})

execute_schedules(schedules, backend, lo_config, **opts)
```
If you have no LO setting, just compose schedule and execute it.
```
# compose schedule
schedule = pulse.Schedule()
#... compose schedule ...
execute_schedules(schedule, backend, **opts)
```

diego-plan9 committed:
Remove underscore from tools and transpiler (#2095)

mtreinish committed:
Add custom gate support to latex circuit drawer (#2049)

* Add custom gate support to latex circuit drawer

This commit adds support to the latex circuit drawer backend to draw
custom gates. Currently when a circuit with a custom element is passed
into the latex circuit drawer it is silently ignored and not drawn. This
supports a custom gate with any number of qubits. This also has a bonus
effect if there was a standard gate which wasn't accounted for in the
latex drawer (we've had bugs like this in the past, see #1880) this will
also fix that for us, assuming the node name matches the expected gate
name (which isn't always the case). The only limitation with this is that
it doesn't support conditionals on the custom gates.

Fixes #1942

* Fix lint

* Handle out of order and noncontiguous custom gates

* Remove stray prints

taalexander commented:
@eggerdj, the rational for reusing a `Schedule` with different Los is to allow frequency sweeps experiments to be easily realized with a minimal sized QObj. For example, CW spectroscopy of a qubit can be performed with a single schedule repeated *n* times with *n* linearly spaced lo frequencies.

taalexander commented on qiskit/compiler/assembler.py:
```diff
@@ -13,9 +13,15 @@
+from qiskit.pulse import ConfiguredSchedule, Snapshot
+from qiskit.pulse.commands import (DriveInstruction, FrameChangeInstruction,
+                                   PersistentValueInstruction, AcquireInstruction)
+from qiskit.qobj import (QasmQobj, PulseQobj, QobjExperimentHeader, QobjHeader,
```
This is just a suggestion, but it may be better to just use `from qiskit import qobj` and then call as `qobj.QobjConditional` rather than pollute the namespace with many explicit Qobj class imports. https://google.github.io/styleguide/pyguide.html#22-imports

taalexander commented on qiskit/compiler/assembler.py:
```diff
@@ -126,3 +132,151 @@ def assemble_circuits(circuits, run_config=None, qobj_header=None, qobj_id=None)
+    default_meas_lo_freq = dict_config.get('meas_lo_freq', None)
+
+    user_pulselib = set()
+    for ii, configed in enumerate(schedules):
```
Could you please use a more explicit variable name than `ii`, is this something like `exp_idx`? Also, what is the meaning of `configed`? Shorthand for `configured`?

taalexander commented on qiskit/compiler/assembler.py:
```diff
@@ -126,3 +132,151 @@ def assemble_circuits(circuits, run_config=None, qobj_header=None, qobj_id=None)
+
+        # use LO frequency configs
+        lo_freqs = {}
+        if configed.config and configed.config.user_lo_dic:
```
Should this be `user_lo_dict`?

taalexander commented on qiskit/compiler/assembler.py:
```diff
@@ -126,3 +132,151 @@ def assemble_circuits(circuits, run_config=None, qobj_header=None, qobj_id=None)
+        commands = []
+        for block in schedule.flat_instruction_sequence():
+            pulse_instr = block.instruction
+            if isinstance(pulse_instr, DriveInstruction):
```
Could you please break these into separate method calls? I think this would make it easier to understand at a glance. 

It might even make sense to have these as a method for a `Command`, `to_qobj`. Which would allow users to create their custom `Command`s and submit them to the backend. However, I think this fails for the case of the acquire instruction as the acquire requires information about all other acquires at that time. I think this is more a failing of how the spec defines acquires then anything else.

nkanazawa1989 replied:
`to_qobj` sounds nice, but I want import qobj models only in `assembler` as circuit is.

I believe it works.
```
current_commands  = PulseQobjInstruction(**pulse_instr.to_dict())
```

`AcquireInstruction` has a list of qubit and memory/register slots, and we can generate qobj instruction from a single command instruction. Violation of meas_map can be checked by another validator.

For acquire:
```
current_commands  = PulseQobjInstruction(**pulse_instr.to_dict(meas_level))
```
because application of discriminator, register and kernel depends on meas_level.

taalexander replied:
How are you currently making sure all acquires are present for a qubit in a measurement map? Adding the acquires to the schedule explicitly before assembly?

nkanazawa1989 replied:
https://github.com/nkanazawa1989/qiskit-terra/blob/ada4f464c2a60bbad990d869f3189bc084470137/qiskit/pulse/commands/acquire.py#L89

Currently single acquire command accepts a list of qubit and slots without validation. I'm writing a schedule validator and this will be added in another PR. This validator will check meas_map.

```
sched_validator = PulseValidator(backend)

sched_validator(schedule) # -> bool, warnings, erros
```

taalexander replied:
I guess what I mean is how do we enforce that all acquires are present in a measurement map for our current examples? Are you just making sure you add acquires for every qubit?

nkanazawa1989 replied:
I believe schedule is independent of device constraints, it should be able to create arbitrary sequence. All violations of the backend will be evaluated before it goes to the backend.
Another idea is to change `qubit` argumanent of `AcquisitionInstruction` to `meas_map` itself. `DeviceSpecification` returns a list of `meas_map` and a user specifies qubits by a unit of `meas_map`. In this case, we can enforce a user to align with `meas_map`.

taalexander commented on qiskit/compiler/assembler.py:
```diff
@@ -126,3 +132,151 @@ def assemble_circuits(circuits, run_config=None, qobj_header=None, qobj_id=None)
+                                               config=experimentconfig))
+
+    # generate qobj pulse library
+    qobj_default_pulselib = list(map(lambda p:
```
Could we use a list comprehension here rather than a `list` and `map` call?

taalexander commented on qiskit/compiler/assembler.py:
```diff
@@ -126,3 +132,151 @@ def assemble_circuits(circuits, run_config=None, qobj_header=None, qobj_id=None)
+                                     QobjPulseLibrary(name=p['name'], samples=p['samples']),
+                                     dict_config.get('pulse_library', []))
+                                 )
+    qobj_user_pulselib = list(map(lambda p:
```
same as above.

taalexander commented on qiskit/test/reference_pulses.py:
```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
```
You could rebase on my pulse library PR if would like to use the tested pulses within that.

taalexander commented on test/python/compiler/test_assembler.py:
```diff
@@ -110,6 +114,28 @@ def test_assemble_opaque_inst(self):
         self.assertEqual(qobj.experiments[0].instructions[0].memory, [3, 0])
         self.assertEqual(qobj.experiments[0].instructions[0].params, [0.5, 0.4])
 
+    def test_assemble_single_schedule(self):
```
We will have to add many more test schedules for of edge cases that may be triggered.

taalexander reviewed (changes requested):
Looks good, just a couple small changes.

taalexander committed:
Merge branch 'master' into openpulse-channels

nkanazawa1989 committed:
OpenPulse frontend - channels module (#2096)

* OpenPulse Command Group:

commit 3a4034d0150d18e336106071fa52abaa369619e7
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 17:25:30 2019 +0900

    edit CHANGELOG

commit 5a00c0930486abbdab0714f6efddb169669d06d9
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 15:59:26 2019 +0900

    code cleanup

commit 123b110ae8bcc45fad7f49dd58e7653533381e27
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 15:09:04 2019 +0900

    Update plot method

commit a413aa1b4803958135dbc379b58f99f6f4c1befe
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 13:21:20 2019 +0900

    Add test case

commit 9a3136aebb95c9ccc94284df762c9cb375539ad5
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 02:06:55 2019 +0900

    Add pulse validity check to functional_pulse

commit 6915d9cdfc17f7ce86dae972a43346effcebac9e
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 01:56:25 2019 +0900

    Add __init__ to command group and change module names

commit dab9b71606a0faee3d244578460e00b36182f218
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 01:43:11 2019 +0900

    Add acquire, frame change, snapshot, persistent_value

commit ddd6cf60517a76ab0b4e9e2764d7b59533d8b9da
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 01:29:01 2019 +0900

    Add draw method to SamplePulse

commit 07b0b0ad2728818ad8ce7cbd5eced7b80b2788c3
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Mon Feb 11 11:48:33 2019 +0900

    fix module name

commit 8ad8dbdcf565cc494a4809c84cdde24f1336c9c8
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Mon Feb 11 10:40:03 2019 +0900

    move functional pulse to command module

commit 2e8a0c5bd55ec0725cb80787aaeb8f2a7f8fad73
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 14:45:50 2019 +0900

    correct module name

commit ab1699eb7efcec25d8987add792757b2a0224225
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 14:31:32 2019 +0900

    reformat code

commit 07e13f08d495897cc710786ef22a3d44bbc8e76f
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 13:27:44 2019 +0900

    add test cases for functional pulse

commit f72f666b9957c0263c2c829de465ed7ec7aa43f3
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 12:57:50 2019 +0900

    add savefig option in plot()

commit a6b34ac2d064fd8fc2e5e58ba187aebcb38cb80c
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 12:07:14 2019 +0900

    allow integer value input to tolist()

commit 6367e3ebfc2612594b5432c247442a944194f7f3
Merge: 6d5f824f ccccd07e
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 06:44:17 2019 +0900

    Merge remote-tracking branch 'origin/pulse_class' into pulse_class

commit 6d5f824faa9846a6e346536efd5f190b99a2a86b
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 06:42:47 2019 +0900

    remove scipy requirement from setup.py

commit ccccd07ea1fb1102b74a71d5472ee320586ae79f
Merge: f7335059 f239ff9a
Author: Ali Javadi-Abhari <ajavadia@users.noreply.github.com>
Date:   Tue Jan 22 16:39:14 2019 -0500

    Merge branch 'master' into pulse_class

commit f7335059d75c39c1627fb9f0394ba90df476730f
Author: nkanazawa <39517270+nkanazawa1989@users.noreply.github.com>
Date:   Tue Jan 22 14:39:32 2019 +0900

    Update CHANGELOG.rst

commit fb36a13b6033ef84d18570a1bc055156b5996f6a
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Jan 22 13:39:04 2019 +0900

    add legend and include import in plot()

commit e5cf271030556ffe717b1bae2ce8934ce4b72160
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Jan 22 01:45:25 2019 +0900

    optimize import

commit 667e4fda7d8fc10d3d08fcbf186fdfa1ee4e20d8
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Jan 22 01:28:08 2019 +0900

    separate visualization, use callables instead of str parse

commit 07173bd897688809239ea7b28dcf1ceaedca3f01
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 17:17:03 2019 +0900

    param -> params

commit cfdcde40648f5a561b380789741cfb4f6243a55c
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 14:08:12 2019 +0900

    correct copyright

commit 3121c572a22669f235e5c4bc0729284d456ef70f
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 13:26:55 2019 +0900

    reformat code

commit 8668c882d57aa39ef753310d0de21a77f5730a98
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 09:20:47 2019 +0900

    reformat code

commit 472b585ea0378e56c29ed915eabcd2a83511da4b
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 07:22:31 2019 +0900

    add pulse class to manage pulse envelope

* Remove insignificant pylint disables

* update pulse_drawer

* Update qiskit/pulse/commands/_functional_pulse.py

remove abbreviation

Co-Authored-By: nkanazawa1989 <39517270+nkanazawa1989@users.noreply.github.com>

* remove abbreviation

* Reformat docstring and parameter names

* import PulseCommand and SamplePulse in __init__

* remove memory/register slot arguments

* add name to all commands

* add t0 to pulse command

* change import of superclass and remove leading underscores

* minor revisions

* fix xrange when interpolated

* remove timing information from command gruop

* initial commit (channels)

* factoring a bit

* WIP: add pulse_schedule

* WIP: add minimal interfaces

* minimal implementation

* fix typo

* change Schedule a it

* WIP: refactor channels

* fix supported commands

* add lo_frequency as member

* add exceptions for pulse

* add channel factory

* add command library

* WIP execute initial commit

* WIP complete qobj converter flow

* exclude backend embedding from schedule_to_qobj

* Add data structure of PulseCommands to generate Qobj

* add superclass of measurement options

* refactor channels

* fix bug of sample pulse without name

* fix bug of __eq__ in sample pulse

* fix a bug

* refactor add() and add name

* refactor channels

* add exceptions for pulse

* remove pylint error

* fix bugs

* fix a bug in channel name

* expose channel factory

* refactor channel factory

* fix bugs on naming

* fix overlap check bug

* remove pulse_lib validation

* change command library from set to list

* fix a bug

* fix config bug

* add pulses to lib only when it is SamplePulses

* simplify schedule

* lint

* refactor a bit

* add channel register

* fix bugs

* add channel memory

* refactor channel factory

* debug

* refactor schedule

* wip: look for better memory interface

* rename to channel bank

* refactor channel register

* refactor channels

* refactor channel registers

* fix bugs in channels

* add getters

* fix schedule to follow channel refactoring

* improve error message

* add interface to get lo_freqs

* improve logging

* add backend config to channel bank

* fix backend import bug

* remove name from channel register

* fill channel bank constructor

* fix style errors

* explicit import

* remove supported channel definition from subclasses

* remove config from compiler argument

* fix raises in schedule

* change method name

* get lo_frequencies from channel bank

* remove frequency embedding at compiler

* explicit import

* avoid setting useless kernel/discriminator

* add setter for lo

* raise error when schedule fails to add

* rename channel bank->store and change its interface

* rename ChannelRegister to ChannelList

* add PulseSchedule.append()

* add pulse and QASM models to qiskit.qobj

* add PulseQobj and QASMQobj to qiskit.qobj

* add interface of assemble_schedules

* add unit test

* resolve errors

* remove duplicated dict key in unittest

* modify imports

* PulseQobj/QASMQobj inherit from Qobj

* move qobj generation part to another PR

* add device_spec replacing channel_store

* remove channel_list

* add __eq__ to device_spec

* replace FunctionalPulse class by @function decorator

* rename add->insert, PulseSchedule->Schedule

* avoid name to be required

* remove unused import

* update init of qiskit.pulse

* remove unused FunctionalPulse

* update test for decorator

* a few tweaks

* change interface of Schedule.insert()

* remove FunctionalPulse

* fix imports

* fix pulse.__init__

* fix a bug in append()

* remove name from function

* remove name check from unittest

* add mem_slots, reg_slots and remove snapshot_channel

* add custom validator for measurement options

* restore snapshot channel

* remove n_registers==n_qubits assumption

* add instruction, header, config to BaseQobjExperimentSchema

* update class names (remove Base- prefix, and QASM > Qasm)

* remove PulseQobjHeader, PulseQobjExperimentHeader, QasmQobjHeader, QasmQobjExperimentHeader,

* separate models into specific modules

* Add minor changes

* avoid cyclic import

* add minimal tests for channels

* change prefix of register 'r'->'c'

* lint

* add change log

* more linting

* add assemble_schedule and execute_schedule

* rename PulseChannel->Channel

* lint (temporal)

* replace dict overwrite by .update

* add DeviceSpecification to pulse.__init__

* fix a bug when u_channels==0

* add test for fixing the u_channel==0 case bug completely

* fix #15

* assign unique name for all commands

* assemble commands from class type

* remove unused counter from SamplePulse

* update unittest

* add decorator

* remove all changes except for channels

* lint

* make mem_slots independent of qubits

ajavadia committed:
apply initialize to register for backward compatibility (#2076)

* backward compat apply initialize to register

* lint

* add tests

* lint

eggerdj commented:
@taalexander and @itoko thanks that clarifies. So this means that the RF sources are allowed to "reinitialize" between schedules. I.e. when comparing with the qc_code a schedule is like an inner loop and running several schedules is like running an outer loop?

mtreinish committed:
Fix typo in setuptools extras declaration (#2100)

The name of the key is extras_require not extra_requires to declare
optional dependencies. [1] This commit corrects the typo so it works
correctly.

[1] https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies

itoko committed:
refactor (simplify) schedule module

Commit:
Merge branch 'master' of github.com:Qiskit/qiskit-terra into openpulse-channels

Commit:
Merge branch 'openpulse-commands-rework' of github.com:nkanazawa1989/qiskit-terra into openpulse-commands-rework

itoko committed:
Merge branch 'openpulse-schedule' into openpulse-assembler

itoko committed:
add command property to instruction for assemler

itoko committed:
add command property to instruction for assemler

itoko committed:
add command property to instruction for assemler

Commit:
fix lint

itoko committed:
add runable assembler/execute for schedule

itoko committed:
add runable assembler/execute for schedule

itoko committed:
debug test

itoko committed:
debug test

itoko committed:
restore accidentally deleted file

itoko committed:
add a corner case test

itoko committed:
add a corner case test

ajavadia committed:
honor initial_layout if it makes the circuit compatible with coupling_map (#2088)

* change 2q and 3q dag methods to return gate nodes not any node

* preserve manual layout

* add tests to check layout is respected

* changelog

* lint

* lint++

* update noise adaptive layout to use dag.twoQ_gates

* undoing changes in the pickles

itoko commented:
Updated interface:
```
[Proposed (1 schedule : n user_lo_dicts)]
# compose schedule
schedule = pulse.Schedule()
#... compose schedule ...

# set LO frequency
user_lo_dicts=[]
for drive_lo_freq in drive_lo_freqs:
    freq_d0 = freq_d0_default + drive_lo_freq
    user_lo_dicts.append(UserLoDict({device.q[0].drive: freq_d0}))

execute_schedules(schedule, backend, user_lo_dicts, **opts)
```
```
[Proposed (n schedules : 1 user_lo_dict)]
# compose to schedules
schedules = []
for ii in pulses:
    #... compose ii-th schedule ...
    schedules.append(schedule)

# set LO frequency
lo_config = UserLoDict({device.q[0].drive: freq_d0})

execute_schedules(schedules, backend, lo_config, **opts)
```
If you specify n schedules with m user_lo_dicts, you can execute n * m experiments.

nkanazawa1989 commented:
How can we execute below experiments?
```
schedule1: user_lo_dicts1
schedule2: user_lo_dicts1
schedule3: user_lo_dicts1
schedule4: user_lo_dicts2
schedule5: user_lo_dicts2
schedule6: user_lo_dicts2
```

Of course we can execute `schedule1-3` and `schedule4-6` as separated jobs. However, a qubit frequency is drifting due to TLS, and we may want to combine different `user_lo_dict`s and `schedule`s in a single job to minimize the effect of frequency drift as much as possible. This will be a critical issue when the QC have long queue and the interval between jobs is very learge.

I think `schedule` and `use_lo_dict` can be combined outside of `execute` as a user desired.

nkanazawa1989 commented:
> 
> 
> @taalexander and @itoko thanks that clarifies. So this means that the RF sources are allowed to "reinitialize" between schedules. I.e. when comparing with the qc_code a schedule is like an inner loop and running several schedules is like running an outer loop?

I believe users will be able to decide which loop is which.

taalexander committed:
Merge branch 'master' into openpulse-commands-rework-rebased

nkanazawa1989 committed:
OpenPulse frontend - commands module rework (#2104)

* OpenPulse Command Group:

commit 3a4034d0150d18e336106071fa52abaa369619e7
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 17:25:30 2019 +0900

    edit CHANGELOG

commit 5a00c0930486abbdab0714f6efddb169669d06d9
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 15:59:26 2019 +0900

    code cleanup

commit 123b110ae8bcc45fad7f49dd58e7653533381e27
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 15:09:04 2019 +0900

    Update plot method

commit a413aa1b4803958135dbc379b58f99f6f4c1befe
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 13:21:20 2019 +0900

    Add test case

commit 9a3136aebb95c9ccc94284df762c9cb375539ad5
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 02:06:55 2019 +0900

    Add pulse validity check to functional_pulse

commit 6915d9cdfc17f7ce86dae972a43346effcebac9e
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 01:56:25 2019 +0900

    Add __init__ to command group and change module names

commit dab9b71606a0faee3d244578460e00b36182f218
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 01:43:11 2019 +0900

    Add acquire, frame change, snapshot, persistent_value

commit ddd6cf60517a76ab0b4e9e2764d7b59533d8b9da
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 01:29:01 2019 +0900

    Add draw method to SamplePulse

commit 07b0b0ad2728818ad8ce7cbd5eced7b80b2788c3
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Mon Feb 11 11:48:33 2019 +0900

    fix module name

commit 8ad8dbdcf565cc494a4809c84cdde24f1336c9c8
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Mon Feb 11 10:40:03 2019 +0900

    move functional pulse to command module

commit 2e8a0c5bd55ec0725cb80787aaeb8f2a7f8fad73
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 14:45:50 2019 +0900

    correct module name

commit ab1699eb7efcec25d8987add792757b2a0224225
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 14:31:32 2019 +0900

    reformat code

commit 07e13f08d495897cc710786ef22a3d44bbc8e76f
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 13:27:44 2019 +0900

    add test cases for functional pulse

commit f72f666b9957c0263c2c829de465ed7ec7aa43f3
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 12:57:50 2019 +0900

    add savefig option in plot()

commit a6b34ac2d064fd8fc2e5e58ba187aebcb38cb80c
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 12:07:14 2019 +0900

    allow integer value input to tolist()

commit 6367e3ebfc2612594b5432c247442a944194f7f3
Merge: 6d5f824f ccccd07e
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 06:44:17 2019 +0900

    Merge remote-tracking branch 'origin/pulse_class' into pulse_class

commit 6d5f824faa9846a6e346536efd5f190b99a2a86b
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 06:42:47 2019 +0900

    remove scipy requirement from setup.py

commit ccccd07ea1fb1102b74a71d5472ee320586ae79f
Merge: f7335059 f239ff9a
Author: Ali Javadi-Abhari <ajavadia@users.noreply.github.com>
Date:   Tue Jan 22 16:39:14 2019 -0500

    Merge branch 'master' into pulse_class

commit f7335059d75c39c1627fb9f0394ba90df476730f
Author: nkanazawa <39517270+nkanazawa1989@users.noreply.github.com>
Date:   Tue Jan 22 14:39:32 2019 +0900

    Update CHANGELOG.rst

commit fb36a13b6033ef84d18570a1bc055156b5996f6a
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Jan 22 13:39:04 2019 +0900

    add legend and include import in plot()

commit e5cf271030556ffe717b1bae2ce8934ce4b72160
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Jan 22 01:45:25 2019 +0900

    optimize import

commit 667e4fda7d8fc10d3d08fcbf186fdfa1ee4e20d8
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Jan 22 01:28:08 2019 +0900

    separate visualization, use callables instead of str parse

commit 07173bd897688809239ea7b28dcf1ceaedca3f01
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 17:17:03 2019 +0900

    param -> params

commit cfdcde40648f5a561b380789741cfb4f6243a55c
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 14:08:12 2019 +0900

    correct copyright

commit 3121c572a22669f235e5c4bc0729284d456ef70f
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 13:26:55 2019 +0900

    reformat code

commit 8668c882d57aa39ef753310d0de21a77f5730a98
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 09:20:47 2019 +0900

    reformat code

commit 472b585ea0378e56c29ed915eabcd2a83511da4b
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 07:22:31 2019 +0900

    add pulse class to manage pulse envelope

* Remove insignificant pylint disables

* update pulse_drawer

* Update qiskit/pulse/commands/_functional_pulse.py

remove abbreviation

Co-Authored-By: nkanazawa1989 <39517270+nkanazawa1989@users.noreply.github.com>

* remove abbreviation

* Reformat docstring and parameter names

* import PulseCommand and SamplePulse in __init__

* remove memory/register slot arguments

* add name to all commands

* add t0 to pulse command

* change import of superclass and remove leading underscores

* minor revisions

* fix xrange when interpolated

* remove timing information from command gruop

* initial commit (channels)

* factoring a bit

* WIP: add pulse_schedule

* WIP: add minimal interfaces

* minimal implementation

* fix typo

* change Schedule a it

* WIP: refactor channels

* fix supported commands

* add lo_frequency as member

* add exceptions for pulse

* add channel factory

* add command library

* WIP execute initial commit

* WIP complete qobj converter flow

* exclude backend embedding from schedule_to_qobj

* Add data structure of PulseCommands to generate Qobj

* add superclass of measurement options

* refactor channels

* fix bug of sample pulse without name

* fix bug of __eq__ in sample pulse

* fix a bug

* refactor add() and add name

* refactor channels

* add exceptions for pulse

* remove pylint error

* fix bugs

* fix a bug in channel name

* expose channel factory

* refactor channel factory

* fix bugs on naming

* fix overlap check bug

* remove pulse_lib validation

* change command library from set to list

* fix a bug

* fix config bug

* add pulses to lib only when it is SamplePulses

* simplify schedule

* lint

* refactor a bit

* add channel register

* fix bugs

* add channel memory

* refactor channel factory

* debug

* refactor schedule

* wip: look for better memory interface

* rename to channel bank

* refactor channel register

* refactor channels

* refactor channel registers

* fix bugs in channels

* add getters

* fix schedule to follow channel refactoring

* improve error message

* add interface to get lo_freqs

* improve logging

* add backend config to channel bank

* fix backend import bug

* remove name from channel register

* fill channel bank constructor

* fix style errors

* explicit import

* remove supported channel definition from subclasses

* remove config from compiler argument

* fix raises in schedule

* change method name

* get lo_frequencies from channel bank

* remove frequency embedding at compiler

* explicit import

* avoid setting useless kernel/discriminator

* add setter for lo

* raise error when schedule fails to add

* rename channel bank->store and change its interface

* rename ChannelRegister to ChannelList

* add PulseSchedule.append()

* add pulse and QASM models to qiskit.qobj

* add PulseQobj and QASMQobj to qiskit.qobj

* add interface of assemble_schedules

* add unit test

* resolve errors

* remove duplicated dict key in unittest

* modify imports

* PulseQobj/QASMQobj inherit from Qobj

* move qobj generation part to another PR

* add device_spec replacing channel_store

* remove channel_list

* add __eq__ to device_spec

* replace FunctionalPulse class by @function decorator

* rename add->insert, PulseSchedule->Schedule

* avoid name to be required

* remove unused import

* update init of qiskit.pulse

* remove unused FunctionalPulse

* update test for decorator

* a few tweaks

* change interface of Schedule.insert()

* remove FunctionalPulse

* fix imports

* fix pulse.__init__

* fix a bug in append()

* remove name from function

* remove name check from unittest

* add mem_slots, reg_slots and remove snapshot_channel

* add custom validator for measurement options

* restore snapshot channel

* remove n_registers==n_qubits assumption

* add instruction, header, config to BaseQobjExperimentSchema

* update class names (remove Base- prefix, and QASM > Qasm)

* remove PulseQobjHeader, PulseQobjExperimentHeader, QasmQobjHeader, QasmQobjExperimentHeader,

* separate models into specific modules

* Add minor changes

* avoid cyclic import

* add minimal tests for channels

* change prefix of register 'r'->'c'

* lint

* add change log

* more linting

* make commands callable

* make channels hashable

* add working example code of Schedule

* add assemble_schedule and execute_schedule

* reorder class defs and add prints

* simplify Schedule

* rename PulseChannel->Channel

* lint (temporal)

* replace dict overwrite by .update

* rename PulseChannel->Channel

* lint

* add getters for qobj creation

* simplify snapshot command

* rename Pulse -> Instruction

* add DeviceSpecification to pulse.__init__

* fix a bug when u_channels==0

* add test for fixing the u_channel==0 case bug completely

* fix #15

* simplify Instruction

* assign unique name for all commands

* assemble commands from class type

* remove unused counter from SamplePulse

* update unittest

* refactor schedule-related interface

* unify exceptions into PulseError

* add a test

* add decorator

* update assembler

* update lo frequency

* add check length of mem_slots

* remove all changes except for channels

* remove all changes except for commands-rework

* lint

* make mem_slots independent of qubits

* fix a bug in Acquire constructor with default values

* refactor meas_opts

* make sample pulses hashable

* make commands callable

* make channels hashable

* add working example code of Schedule

* reorder class defs and add prints

* simplify Schedule

* rename PulseChannel->Channel

* lint

* add getters for qobj creation

* simplify snapshot command

* rename Pulse -> Instruction

* simplify Instruction

* refactor schedule-related interface

* unify exceptions into PulseError

* add check length of mem_slots

* remove all changes except for commands-rework

* make sample pulses hashable

* fix a bug in Acquire constructor with default values

* refactor meas_opts

* fix lint

eggerdj commented:
This make perfect sense.  But I wonder if we have a small naming issue? Consider the following example. 

I do spectroscopy on the 1-2 transition of a three level system by doing a pi-pulse on 0-1 and then spectroscopy on 1-2. The 1-2 transition is approx 300 MHz detuned from 0-1. My setup has IQ mixing. An RF source generates the LO. This RF source has a certain range of frequencies it can generate (typically 10 KHz to 20 GHz, I would call this the LO range). An AWG generates the pulses and adds a sideband to the pulses (the AWG also has a bandwidth, e.g. +- 350 MHz).  In order to do spectroscopy I have different schedules in which I change the side-band frequency of the second pulse using the AWG by say 250-350 MHz. But in this experiment the LO is always the same.

Does this workout with the current terminology?

taalexander commented:
I believe this still works. In your backend setting a `qubit_lo_freq` would be an actual `LO` (the microwave source). You could then still add a sideband with the AWG by defining the proper pulse. It is up to the backend to determine how it interprets the input pulse QObj and implements the specified experiment.

mtreinish committed:
Test to verify mpl draws an empty circuit empty (#2105)

* Test to verify mpl draws an empty circuit empty

This commit adds a test that verifies that when we call the mpl backend
it draws an empty image of the expected size. This is just a white image
background of the default size.

Related to #1845 and #2097

* Skip if matplotlib is not installed

* Only import from matplotlib if installed

nkanazawa1989 commented:
Yes, that works. You can create two `schedules` for 0-1 and 1-2 transitions and combine them into a list, and create single `user_lo_config`. Then execute the experiment as follows.
```
job = execute_schedule(schedules, backend, user_lo_dicts=user_lo_config, **opt)
```

However we cannot handle experiments such as a calibration of pi pulse and measurement pulse in a single job.

1ucian0 committed:
"swap followed by measure" optimization pass (#1890)

1ucian0 committed:
Performance issue because optimizations-until-fixed-point approach (#2107)

1ucian0 committed:
swap_mapper as a pass (#2094)

* qiskit/mapper/mapping.py moved to qiskit/transpiler/passes/mapping/legacy_swap.py

* LegacySwap added to the mapper meta tests

* layer_permutation params

* swap_mapper_layer_update params

* circuit_graph -> dag

* lint

* seed update

* long line

Commit:
add to_dict to commands

Commit:
add to_dict to commands

Commit:
use list comprehension to generate qobj pulse lib

Commit:
use list comprehension to generate qobj pulse lib

itoko committed:
Merge branch 'openpulse-schedule' of https://github.com/nkanazawa1989/qiskit-terra into openpulse-schedule

itoko committed:
update a test due to change in channels

itoko commented:
> How can we execute below experiments?
> 
> ```
> schedule1: user_lo_dicts1
> schedule2: user_lo_dicts1
> schedule3: user_lo_dicts1
> schedule4: user_lo_dicts2
> schedule5: user_lo_dicts2
> schedule6: user_lo_dicts2
> ```
I had second thought on execute_schedule interface. I think we should go back to the original interface: it can handle N schedules : N user_lo_dicts.
By n schedules : m user_lo_dicts -> n*m experiments design, this example with one job is impossible.
By N schedules : N user_lo_dicts -> N experiments design, user can run this example with one job and also cover n*m experiments case by creating N=n*m schedules and N=n*m user_lo_dicts manually.
This means N schedules : N user_lo_dicts -> N experiments design is more general, i.e. better.
What do you think? > @nkanazawa1989

eggerdj commented:
Thanks for the explanations.

itoko committed:
rename begin/end_time to start/stop_time

itoko committed:
Merge branch 'master' into openpulse-schedule

itoko committed:
add lo_range check

itoko committed:
rename TimeslotOccupancy to TimeslotCollection

itoko committed:
remove __lshift__ syntax sugar (will be readd later)

itoko committed:
fix a bug in type hint

itoko committed:
add tests for UserLoDict

itoko committed:
improve lo_range handling

itoko committed:
add a type check

nkanazawa1989 commented:
I agree @itoko . N:N can exactly explain the setup of experiment.

Commit:
move to_dict functionality to compiler

remove unused import

Commit:
move to_dict functionality to compiler

remove unused import

Commit:
add pulse converter

optimize import

Commit:
add pulse converter

optimize import

Commit:
wip unittest

wip unittest

Commit:
wip unittest

wip unittest

itoko committed:
Merge branch 'openpulse-assembler' of https://github.com/nkanazawa1989/qiskit-terra into openpulse-assembler

Commit:
add converter to argument

Commit:
add converter to argument

Commit:
rename converter

Commit:
rename converter

Commit:
Merge branch 'openpulse-assembler-qobj-rework' of github.com:nkanazawa1989/qiskit-terra into openpulse-assembler-qobj-rework

Commit:
update property name

itoko committed:
change to check None outside

taalexander committed:
Rename test_experiment_condition.py to test_experiment_configurations.py

taalexander committed:
Merge pull request #42 from nkanazawa1989/openpulse-assembler-lo-range

Add LO range check

taalexander committed:
Merge pull request #43 from nkanazawa1989/openpulse-assembler-qobj-rework

Openpulse assembler qobj rework

taalexander commented on qiskit/compiler/assembler.py:
```diff
@@ -126,3 +132,94 @@ def assemble_circuits(circuits, run_config=None, qobj_header=None, qobj_id=None)
+                                               config=experimentconfig))
+
+    # generate qobj pulse library
+    qobj_default_pulselib = [QobjPulseLibrary(name=p['name'], samples=p['samples'])
```
I don't think we should include the default pulses in the pulse library. If they are used in a schedule they will already be included with the user pulses.

taalexander reviewed (changes requested):
Looking good, one minor change.

itoko commented:
Example codes:
```
[Case - No LO settings]
# compose schedule
schedule = pulse.Schedule()
#... compose schedule ...
execute_schedules(schedule, backend, **opts)
```
```
[Case - 1 schedule : n lo_configs]
# compose schedule
schedule = pulse.Schedule()
#... compose schedule ...

# set LO frequency
lo_configs=[]
for drive_lo_freq in drive_lo_freqs:
    freq_d0 = freq_d0_default + drive_lo_freq
    lo_configs.append(UseLoDict({device.q[0].drive: freq_d0}))

execute_schedules(schedule, backend, lo_configs, **opts)
```
```
[Case - n schedules : 1 lo_config]
# compose to schedules
schedules = []
for ii in pulses:
    #... compose ii-th schedule ...
    schedules.append(schedule)

# set LO frequency
lo_config = UseLoDict({device.q[0].drive: freq_d0})

execute_schedules(schedules, backend, lo_config, **opts)
```
```
[Case - n*m schedules : n*m lo_configs from n schedules : m lo_configs]
# compose to n schedules
schedules = []
for ii in pulses:
    #... compose ii-th schedule ...
    schedules.append(schedule)

# m LO frequency settings
lo_configs=[]
for drive_lo_freq in drive_lo_freqs:
    freq_d0 = freq_d0_default + drive_lo_freq
    lo_configs.append(UseLoDict({device.q[0].drive: freq_d0}))

schedules = schedules * len(lo_configs)
lo_configs = lo_configs * len(schedules)
assert(len(schedules) == len(lo_configs))

execute_schedules(schedules, backend, lo_config, **opts)
```
